galaxy_ng/app/migrations/0026_move_perms_to_roles.py

```diff
@@ -0,0 +1,191 @@
+
+
+def move_permissions_to_roles(apps, schema_editor):
+    Group = apps.get_model("galaxy", "Group")
+    GroupRole = apps.get_model("core", "GroupRole")
+    UserRole = apps.get_model("core", "UserRole")
+    Permission = apps.get_model("auth", "Permission")
+    Role = apps.get_model("core", "Role")
+    UserObjectPermission = apps.get_model("guardian", "UserObjectPermission")
+    GroupObjectPermission = apps.get_model("guardian", "GroupObjectPermission")
+
+    group_roles = []
+    user_roles = []
+
+    # Model Permissions
+    for group in Group.objects.all():
+        if group.name == "system:partner-engineers":
+            continue
+        permissions = group.permissions.all()
+        perms_to_remove = []
+
+        # Use Galaxy locked Roles where possible
+        for locked_perm_names, locked_rolename in MODEL_PERMISSION_TRANSLATOR:
+            locked_perms = [Permission.objects.filter(
+                content_type__app_label=app_label, codename=codename
+            ).first() for app_label, codename in locked_perm_names]
+
+            if all(locked_perms):
+                # compare locked role perms to perms
+                for k, locked_perm in enumerate(locked_perms):
+                    if locked_perm in permissions:
+                        # Is that all of the permissions in this locked_role?
+                        if (k + 1) == len(locked_perms):
+                            # add current locked_role to list to add to current group
+                            locked_role, _ = Role.objects.get_or_create(
+                                name=locked_rolename,
+                                defaults={"locked": True}
+                            )
+                            group_roles.append(GroupRole(group=group, role=locked_role))
+
+                            # add permissions from locked_perms to list to be removed
+                            perms_to_remove.extend(locked_perms)
+                        continue
+                    else:
+                        break
+
+            # Handle batches
+            if len(group_roles) > 1000:
+                GroupRole.objects.bulk_create(group_roles)
+                group_roles.clear()
+
+        # remove permissions covered by locked_roles from the groups permission list
+        permissions = remove_locked_perms_from_permissions(perms_to_remove, permissions)
+
+        # Create custom role for current Group with permissions not in Galaxy locked roles
+        if len(permissions) > 0:
+            role, _ = Role.objects.get_or_create(name=f"{group.name}_role")
+            role.permissions.set(permissions)
+            group_roles.append(GroupRole(group=group, role=role))
+
+    # Group Object Permissions
+    for gop in GroupObjectPermission.objects.all():
+        # create GroupRole objects, assign to group from gop
+        role, _ = Role.objects.get_or_create(name=f"{gop.group.name}_role")
+        role.permissions.add(gop.permission)
```

Comment:
That sounds dangerous. You may hit the same group you created before without or for another object, so now you are adding the new permissions to that role making it effective for that other object or even globally.
I think you need to extend the role name with the objects pk to make it unique.